util: fix isBuffer

brendanashworth opened the pull request:
Importing Buffer caused a circular dependency issue. We need to use the
globally exported variable to get around this.

Fixes: https://github.com/nodejs/io.js/issues/1987

This is just quick patch to fix the bug, I hope we can later find a better solution.

brendanashworth commented:
*sigh*, this spurs a linter error:
```
lib/util.js
  665:24  error  Use const Buffer = require('buffer').Buffer; at the beginning of this file  require-buffer

✖ 1 problem (1 error, 0 warnings)
```

cc @silverwind

targos commented:
You can also put [this line](https://github.com/nodejs/io.js/blob/master/lib/buffer.js#L13) before the requires

brendanashworth commented:
Still hacky, but better! Thanks @targos, I switched the order of exports.

I tried to remove the `isBuffer` call altogether (keep it DRY with `buffer.isBuffer`), but that causes another circular dependency. I think the best way forward is to move `util.deprecate` to an internal module.

brendanashworth commented:
Alright, force pushed a fix that moves `deprecate()` to an internal module, allowing us to circumvent the whole thing. Right now, the internal module `.deprecate` is still exposed through `util.deprecate`.

silverwind commented:
@brendanashworth https://github.com/nodejs/io.js/pull/1892 should fix that eventually

As for the linter error, you could just put `/* eslint-disable require-buffer */` in worst case.

silverwind commented:
Also I wonder why no test has caught this circular dependency. Maybe add a small test for `util.isBuffer`?

brendanashworth commented:
@silverwind hmm, looking at #1892, it would reintroduce the dep issue because the internal module references back to the `util` module. Do you think I should just add the linter error hack for now, have #1892 fix the cyclic issue, and call it a day? And I added a small true / false test, I too was surprised it wasn't caught by the test suite.

Added a small commit, summary:
```javascript
> util.isBuffer === Buffer.isBuffer
true
```

silverwind commented:
I'd be fine with a hack if you denote it as such. https://github.com/nodejs/io.js/pull/1892 still needs a bit of work.

brendanashworth commented:
I'd prefer to go through with these commits, without introducing a hack / linter comment, to fix the bug. #1892 shouldn't be hard to port the changes through, as the location of the code is all that changes.

silverwind commented:
True, shouldn't be too hard to have seperate methods later.

LGTM, started off a CI just to be sure:

https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/17/

silverwind commented:
CI failures are unrelated. @brendanashworth can you land it?

This and https://github.com/nodejs/io.js/commit/671e64ac732532892e543c0a8d6efd0b8639e757 are pretty important fixes and we should probably do a quick patch release for them.

cc: @rvagg @Fishrock123 @chrisdickinson

brendanashworth commented:
@silverwind yes, landing now - https://github.com/nodejs/io.js/commit/3806d875d3dbff96041638299a1974579e9a5b4a seems to be important too, [they were talking about a CVE](https://github.com/nodejs/io.js/pull/1811#issuecomment-110793722).

brendanashworth commented:
Thank you, I've merged these commits in https://github.com/nodejs/io.js/compare/671e64ac732532892e543c0a8d6efd0b8639e757...626432d843aaf3abc60b2db2b265b95023fd5938, in time for https://github.com/nodejs/io.js/issues/1996.
